### PR TITLE
Viite 1535 separate linear location

### PR DESCRIPTION
**Pull request by leirasf**

### CODE REVIEW

File: digiroad2-api-viite/src/main/scala/fi/liikennevirasto/digiroad2/ViiteApi.scala [hunk 1]
```diff
@@ -939,24 +957,18 @@ class ViiteApi(val roadLinkService: RoadLinkService, val vVHClient: VVHClient,
...
   private def roadAddressLinkLikeToApi(roadAddressLink: RoadAddressLinkLike): Map[String, Any] = {
-    val roadLinkTypeValue = roadAddressLink match {
-      case ra: RoadAddressLink => if (ra.floating) RoadLinkType.FloatingRoadLinkType.value else roadAddressLink.roadLinkType.value
-      case _ => roadAddressLink.roadLinkType.value
-    }
     Map(
       "success" -> true,
-      "segmentId" -> roadAddressLink.id,
-      "id" -> roadAddressLink.id,
+      "roadAddressId" -> roadAddressLink.id, // TODO Should this be renamed to roadwayId?
```
Comment (Campanha) [inline digiroad2-api-viite/src/main/scala/fi/liikennevirasto/digiroad2/ViiteApi.scala]: rename var to roadwayId and remove TODO?

File: digiroad2-api-viite/src/main/scala/fi/liikennevirasto/digiroad2/ViiteApi.scala [hunk 2]
```diff
@@ -1000,7 +1010,7 @@ class ViiteApi(val roadLinkService: RoadLinkService, val vVHClient: VVHClient,
 
   def roadAddressErrorsToApi(addressError: AddressErrorDetails): Map[String, Long] = {
     Map(
-      "id" -> addressError.id,
+      "id" -> addressError.linearLocationId,
```
Comment (Campanha) [inline digiroad2-api-viite/src/main/scala/fi/liikennevirasto/digiroad2/ViiteApi.scala]: Map("linearLocationId" -> ?

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/util/AssetDataImporter.scala [hunk 3]
```diff
@@ -116,235 +117,189 @@ class AssetDataImporter {
...
-    }
-
-  }
-
   def splitRoadAddresses(roadAddress: RoadAddress, addrMToSplit: Long, roadTypeBefore: RoadType, roadTypeAfter: RoadType, elyCode: Long): Seq[RoadAddress] = {
     // mValue at split point on a TowardsDigitizing road address:
     val splitMValue = roadAddress.startMValue + (roadAddress.endMValue - roadAddress.startMValue) / (roadAddress.endAddrMValue - roadAddress.startAddrMValue) * (addrMToSplit - roadAddress.startAddrMValue)
-    println(s"Splitting road address id = ${roadAddress.id}, tie = ${roadAddress.roadNumber} and aosa = ${roadAddress.roadPartNumber}, on AddrMValue = $addrMToSplit")
-    val roadAddressA = roadAddress.copy(id = fi.liikennevirasto.viite.NewRoadAddress, roadType = roadTypeBefore, endAddrMValue = addrMToSplit, startMValue = if (roadAddress.sideCode == SideCode.AgainstDigitizing)
+    println(s"Splitting roadway id = ${roadAddress.id}, tie = ${roadAddress.roadNumber} and aosa = ${roadAddress.roadPartNumber}, on AddrMValue = $addrMToSplit")
```
Comment (Campanha) [inline digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/util/AssetDataImporter.scala]: should we rename .id to to .roadwayId?

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/util/RoadAddressImporter.scala [hunk 4]
```diff
@@ -130,101 +166,168 @@ class RoadAddressImporter(conversionDatabase: DatabaseDef, vvhClient: VVHClient,
...
+      case (key, addresses) =>
+        addresses.foreach{
+          //add current linear locations
+          add =>
+            val converted = add._1
+            val roadLink = mappedRoadLinks.getOrElse(converted.linkId, mappedHistoryRoadLinks(converted.linkId))
+
+            val linearLocation = adjustLinearLocation(IncomingLinearLocation(converted.roadwayNumber, add._2, converted.linkId, converted.startM, converted.endM, converted.sideCode, getStartCalibrationPointValue(converted), getEndCalibrationPointValue(converted),
+              roadLink.linkSource, FloatingReason.NoFloating, createdBy = "import", converted.x1, converted.y1, converted.x2, converted.y2, converted.validFrom, None), groupedLinkCoeffs(converted.linkId))
+            if(add._1.directionFlag == 1){
```
Comment (Campanha) [inline digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/util/RoadAddressImporter.scala]: what is converted.directionFlag == 1 comparing? The geometry digitizing?

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/AddressLinkBuilder.scala [hunk 5]
```diff
@@ -125,16 +125,17 @@ trait AddressLinkBuilder {
   }
 
   def fuseRoadAddress(roadAddresses: Seq[RoadAddress]): Seq[RoadAddress] = {
```
Comment (leirasf) [inline digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/AddressLinkBuilder.scala]: Add comment to do it on the apply changes user story

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/dao/RoadNetworkDAO.scala [hunk 6]
```diff
@@ -9,47 +9,50 @@ import fi.liikennevirasto.viite.AddressConsistencyValidator.AddressError
 import slick.jdbc.StaticQuery.interpolation
 import slick.jdbc.{StaticQuery => Q}
 
-case class RoadNetworkError(id: Long, roadAddressId: Long, error: AddressError, error_timestamp: Long, network_version: Option[Long])
+case class RoadNetworkError(id: Long, roadwayId: Long, error: AddressError, error_timestamp: Long, network_version: Option[Long])
```
Comment (leirasf) [inline digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/dao/RoadNetworkDAO.scala]: Add the `linearLocationId` field

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/util/RoadAddressImporter.scala [hunk 7]
```diff
@@ -130,101 +166,168 @@ class RoadAddressImporter(conversionDatabase: DatabaseDef, vvhClient: VVHClient,
...
-      case(ra) =>
-        mappedRoadLinks.getOrElse(ra.linkId, Seq()).headOption match {
-          case Some(rl) =>
-            Some(IncomingLrmPosition(ra.lrmId, ra.linkId, ra.startM, ra.endM, ra.sideCode, rl.linkSource, ra.roadwayId))
-          case _ =>
-            mappedHistoryRoadLinks.get(ra.linkId) match {
-              case Some(rl) =>
-                Some(IncomingLrmPosition(ra.lrmId, ra.linkId, ra.startM, ra.endM, ra.sideCode, LinkGeomSource.HistoryLinkInterface, ra.roadwayId))
-              case _ => None
+    //TODO - insert expiredConversionAddresses and historyConversionAddresses
```
Comment (leirasf) [inline digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/util/RoadAddressImporter.scala]: I think this TODO can be deleted.

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/util/RoadAddressImporter.scala [hunk 8]
```diff
@@ -130,101 +166,168 @@ class RoadAddressImporter(conversionDatabase: DatabaseDef, vvhClient: VVHClient,
...
+    historyMappedConversionAddresses.mapValues{
+      case address =>
+        address.sortBy(_.startAddressM).zip(1 to address.size)
+    }.foreach{
+      case (key, addresses) =>
+        val minAddress = addresses.head._1
+        val maxAddress = addresses.last._1
+        val linkIds = addresses.map(_._1.linkId)
+        val currentAddresses = currentConversionAddresses.filter(add => add.roadwayNumber == minAddress.roadwayNumber && linkIds.contains(add.linkId)).sortBy(_.startAddressM)
+        val isReversed = if(currentAddresses.head.linkId == minAddress.linkId && currentAddresses.head.startM == minAddress.startM) 1 else 0
```
Comment (Campanha) [inline digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/util/RoadAddressImporter.scala]: is the isReserved logic correct? or should it be "0 else 1"?

File: viite-UI/src/view/ProjectLinkLayer.js [hunk 9]
```diff
@@ -20,7 +20,7 @@
     var isNotEditingData = true;
     var isActiveLayer = false;
 
-    var styler = new Styler();
+    //var styler = new Styler();
```
Comment (Campanha) [inline viite-UI/src/view/ProjectLinkLayer.js]: styler can be removed?

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/RoadLinkService.scala [hunk 10]
```diff
@@ -226,27 +216,36 @@ class RoadLinkService(val vvhClient: VVHClient, val eventbus: DigiroadEventBus,
...
-      else
-        vvhClient.roadLinkData.fetchByRoadNumbersBoundsAndMunicipalitiesF(bounds, municipalities, roadNumbers, publicRoads),
+    val links = Await.result(vvhClient.roadLinkData.fetchByRoadNumbersBoundsAndMunicipalitiesF(bounds, municipalities, roadNumbers, publicRoads),
       atMost = Duration.Inf)
-
     (enrichRoadLinksFromVVH(links), Seq())._1
   }
 
   def getRoadLinksFromVVH(bounds: BoundingRectangle, roadNumbers: Seq[(Int, Int)], municipalities: Set[Int],
-                          everything: Boolean, publicRoads: Boolean, frozenTimeVVHAPIServiceEnabled: Boolean): Seq[RoadLink] =
```
Comment (plourenco90) [inline digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/RoadLinkService.scala]: we can also remove the frozen from the properties file
digiroad2.VVHRoadlink.frozen=false

File: viite-UI/src/model/SelectedLinkProperty.js [hunk 11]
```diff
@@ -101,10 +100,13 @@
...
         var roadPartNumbers = {roadPartNumber: extractUniqueValues(selectedData, 'roadPartNumber')};
         var elyCodes = {elyCode: extractUniqueValues(selectedData, 'elyCode')};
-        var trackCode = {trackCode: extractUniqueValues(selectedData, 'trackCode')};
+
+        // TODO Check that merge was done correctly
         var discontinuity = {discontinuity: parseInt(extractUniqueValues([endRoadOnSelection], 'discontinuity'))};
-        var startAddressM = {startAddressM: parseInt(extractUniqueValues([startRoadOnSelection], 'startAddressM'))};
-        var endAddressM =   {endAddressM:  parseInt(extractUniqueValues([endRoadOnSelection], 'endAddressM'))};
+        //var startAddressM = {startAddressM: parseInt(extractUniqueValues([startRoadOnSelection], 'startAddressM'))};
+        //var endAddressM =   {endAddressM:  parseInt(extractUniqueValues([endRoadOnSelection], 'endAddressM'))};
```
Comment (plourenco90) [inline viite-UI/src/model/SelectedLinkProperty.js]: I think that, these can be removed

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/dao/LinearLocationDAO.scala [hunk 12]
```diff
@@ -0,0 +1,731 @@
...
+  }
+
+  private def queryList(query: String): List[LinearLocation] = {
+    Q.queryNA[LinearLocation](query).list.groupBy(_.id).map {
+      case (_, list) =>
+        list.head
+    }.toList
+  }
+
+  def fetchById(id: Long): Option[LinearLocation] = {
```
Comment (hestrompa) [inline digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/dao/LinearLocationDAO.scala]: Should we update the names of the methods?
Right now we have "fetchById" and "queryById", and they do the same operation, but one for a single id and the other for a list of ids. Should we keep different names? We have more cases like this one

File: digiroad2-oracle/src/main/resources/db/migration/V0_2__create_tables.sql [hunk 13]
```diff
@@ -1,455 +1,279 @@
...
+  NAME VARCHAR2(32 CHAR) NOT NULL, 
+  ELY NUMBER(2,0), 
+  CREATED_BY VARCHAR2(32) NOT NULL, 
+  CREATED_DATE DATE DEFAULT SYSDATE NOT NULL, 
+  MODIFIED_BY VARCHAR2(32), 
+  MODIFIED_DATE DATE, 
+  ADD_INFO VARCHAR2(1000), 
+  START_DATE DATE, 
+  STATUS_INFO VARCHAR2(1000), 
+  CHECK_COUNTER NUMBER(38,0) DEFAULT 0 NOT NULL,
```
[the remark this answers is not in the document]
Comment (skosonen) [reply digiroad2-oracle/src/main/resources/db/migration/V0_2__create_tables.sql]: CHECK_COUNTER removed.

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/util/AssetDataImporter.scala (hunk 3 above)
Comment (skosonen) [reply digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/util/AssetDataImporter.scala]: If we change the BaseRoadAddress.id to roadwayId, we have a conflict with the existing ProjectLink.roadwayId. Maybe better to see this later?

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/AddressLinkBuilder.scala (hunk 5 above)
Comment (skosonen) [reply digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/AddressLinkBuilder.scala]: Comment added. (This pull request is already so big that Github doesn't show the latest changes any more)

**Pull request merged**